src/engine.ml

```diff
@@ -648,26 +694,27 @@ let timer state =
   let s''' = maybe_drop_lame_duck s'' in
   s''', out @ out'
 
-let incoming_data err (ctx : keys) compress data =
+let incoming_data ?(ts = false) err (ctx : keys) compress data =
   (* ok, from the spec: hmac(explicit iv, encrypted envelope) ++ explicit iv ++ encrypted envelope *)
   let hmac, data = Cstruct.split data Packet.hmac_len in
   let hmac' = Mirage_crypto.Hash.SHA1.hmac ~key:ctx.their_hmac data in
   guard (Cstruct.equal hmac hmac') (err hmac') >>= fun () ->
   let iv, data = Cstruct.split data 16 in
   let dec = Mirage_crypto.Cipher_block.AES.CBC.decrypt ~key:ctx.their_key ~iv data in
   (* dec is: uint32 packet id followed by (lzo-compressed) data and padding *)
-  let hdr_len = 4 + if compress then 1 else 0 in
+  let hdr_len = 4 + (if compress then 1 else 0) + (if ts then 4 else 0) in
```

Comment:
```suggestion
  let hdr_len = 4 + (if ts then 4 else 0) + (if compress then 1 else 0) in
```
Since the timestamp comes before the compression tag, this ordering makes more sense to my brain :)